### PR TITLE
feat(kuma-cp) service insights

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

Same concept, API, and CLI as for MeshInsight. ServiceInsight entity exists 1 to 1 with Meshes. Aggregates services that are implemented by dataplanes in the mesh.
- GET `/service-insights`:
```json
{
 "total": 1,
 "items": [
  {
   "type": "ServiceInsight",
   "name": "default",
   "creationTime": "2020-11-17T01:47:03Z",
   "modificationTime": "2020-11-17T01:47:03Z",
   "lastSync": "2020-11-17T01:47:58.992415979Z",
   "services": {
    "frontend_kuma-demo_svc_8080": {
     "total": 1,
     "online": 1
    },
    "postgres_kuma-demo_svc_5432": {
     "total": 1,
     "online": 1
    },
    "redis_kuma-demo_svc_6379": {
     "total": 1,
     "online": 1
    }
   }
  }
 ],
 "next": null
}
```
GET `/meshes/default/service-insights`:
```json
{
 "type": "ServiceInsight",
 "name": "default",
 "creationTime": "2020-11-17T01:47:03Z",
 "modificationTime": "2020-11-17T01:47:03Z",
 "lastSync": "2020-11-17T01:47:58.992415979Z",
 "services": {
  "frontend_kuma-demo_svc_8080": {
   "total": 1,
   "online": 1
  },
  "postgres_kuma-demo_svc_5432": {
   "total": 1,
   "online": 1
  },
  "redis_kuma-demo_svc_6379": {
   "total": 1,
   "online": 1
  }
 }
}
```
GET `/meshes/default/service-insights/frontend_kuma-demo_svc_8080`:
```json
{
  "total": 1,
  "online": 1
}
```

New command for `kumactl`:
```sh
$ kumactl inspect services --mesh mymesh
SERVICE                       DATAPLANES
frontend_kuma-demo_svc_8080   4/4
postgres_kuma-demo_svc_5432   1/1
redis_kuma-demo_svc_6379      1/1
```
### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### CODE REVIEW

File: pkg/api-server/definitions/service_insight.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package definitions
+
+import (
+	"github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+)
+
+var ServiceInsightWsDefinition = ResourceWsDefinition{
+	Name: "Service Insight",
+	Path: "service-insights",
```
Comment (tharun208) [inline pkg/api-server/definitions/service_insight.go:10]: the pathname should be `service+insights` to be consistent with Dataplane Insights

File: app/kumactl/cmd/inspect/inspect.go [hunk 2]
```diff
@@ -29,5 +29,6 @@ func NewInspectCmd(pctx *kumactl_cmd.RootContext) *cobra.Command {
 	cmd.AddCommand(newInspectDataplanesCmd(ctx))
 	cmd.AddCommand(newInspectZonesCmd(ctx))
 	cmd.AddCommand(newInspectMeshesCmd(ctx))
+	cmd.AddCommand(newInspectServicesCmd(ctx))
```
Comment (tharun208) [inline app/kumactl/cmd/inspect/inspect.go:32]: code completion and `make help` need to be updated

**Review by tharun208 — CHANGES_REQUESTED** [review]

File: app/kumactl/cmd/inspect/inspect_services.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			default:
+				printer, err := printers.NewGenericPrinter(format)
+				if err != nil {
+					return err
+				}
+				return printer.Print(rest_types.From.Resource(insights), cmd.OutOrStdout())
+			}
+		},
+	}
+	cmd.PersistentFlags().StringVarP(&ctx.mesh, "mesh", "m", "default", "mesh")
```
Comment (tharun208) [inline app/kumactl/cmd/inspect/inspect_services.go:55]: can we move this to `inspect.go` and add `mesh` as a field to inspect context?

File: pkg/insights/resyncer.go [hunk 4]
```diff
@@ -119,7 +126,68 @@ func meshScoped(t model.ResourceType) bool {
 	return true
 }
 
-func (p *resyncer) resync() error {
+func (p *resyncer) createOrUpdateServiceInsights() error {
```
Comment (jakubdyszkiewicz) [inline pkg/insights/resyncer.go:129]: introduce mutex so we are not updating from ticker and events at the same time

File: pkg/insights/resyncer.go [hunk 5]
```diff
@@ -119,7 +126,68 @@ func meshScoped(t model.ResourceType) bool {
...
+			}
+		}
+	}
+	for _, stat := range insight.Services {
+		stat.Offline = stat.Total - stat.Online
+	}
+	if err := manager.Upsert(p.rm, model.ResourceKey{Mesh: model.NoMesh, Name: mesh}, &core_mesh.ServiceInsightResource{}, func(resource model.Resource) {
+		insight.LastSync = proto.MustTimestampProto(core.Now())
+		_ = resource.SetSpec(insight)
+	}); err != nil {
```
Comment (jakubdyszkiewicz) [inline pkg/insights/resyncer.go:184]: I think you need to if when Mesh is deleted like I did with the other one

File: pkg/insights/resyncer.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/insights/resyncer.go:132]: thanks, fixed

File: pkg/insights/resyncer.go (hunk 5 above)
Comment (lobkovilya) [reply pkg/insights/resyncer.go]: thanks, fixed

File: pkg/core/resources/apis/mesh/service_insight.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		return nil
+	}
+}
+
+func (m *ServiceInsightResource) Validate() error {
+	return nil
+}
+
+func (m *ServiceInsightResource) Scope() model.ResourceScope {
+	return model.ScopeGlobal
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/service_insight.go:53]: The current implementation might be not really obvious, but it satisfies requirements. We want to be able to receive ServiceInsight for each mesh but at the same time Service as an entity is not mesh-scoped. So one service could be represented by several data planes that belong to the different meshes.
That's why here I calculate 1 ServiceInsight object for 1 Mesh. So it's absolutely the same way as we have for MeshInsight and that's why it also has `ScopeGlobal`

**Review by jakubdyszkiewicz — APPROVED** [review]
looks good, please double check that kumactl works

**Pull request merged**